File: archive/programmers/87694/ppoddo/아이템 줍기.py
from collections import deque
def solution(rectangle, characterX, characterY, itemX, itemY):
    # 좌표 최대 50 → 2배 하면 100
    # 여유 있게 102 크기 사용
    # 0은 빈공간, 1은 테두리, 2는 직사각형 내부
    board = [[0] * 102 for _ in range(102)]

    # 모든 직사각형 좌표를 2배 확대해서 board에 표시
    for x1, y1, x2, y2 in rectangle:
        x1 *= 2
        y1 *= 2
        x2 *= 2
        y2 *= 2

        for x in range(x1, x2 + 1):
            for y in range(y1, y2 + 1):

                # 직사각형 내부
                if x1 < x < x2 and y1 < y < y2:
                    board[x][y] = 2

                # 직사각형 테두리
                elif board[x][y] != 2:
                    board[x][y] = 1

    # 캐릭터와 아이템 좌표도 2배 확대
    start_x = characterX * 2
    start_y = characterY * 2
    target_x = itemX * 2
    target_y = itemY * 2

    # BFS 시작
    queue = deque()
    queue.append((start_x, start_y, 0))

    visited = [[False] * 102 for _ in range(102)]
    visited[start_x][start_y] = True

    directions = [(-1, 0),(1, 0),(0, -1),(0, 1)]

    while queue:
        x, y, distance = queue.popleft()

        # 아이템 도달 시 종료
        if x == target_x and y == target_y:
            return distance // 2

        for dx, dy in directions:
            nx = x + dx
            ny = y + dy

            if nx < 0 or nx >= 102 or ny < 0 or ny >= 102:
                continue

            if visited[nx][ny]:
                continue

            # 테두리만 이동 가능
            if board[nx][ny] != 1:
                continue

            visited[nx][ny] = True
            queue.append((nx, ny, distance + 1))

File: archive/programmers/87694/haeun/아이템 줍기.js
function solution(rectangle, characterX, characterY, itemX, itemY) {
  const dx = [0, 1, 0, -1];
  const dy = [1, 0, -1, 0];

  const borders = Array.from({ length: 102 }, () => Array(102).fill(0));

  // 좌표 내부의 점/면 모두 칠하기
  for (const [lx, ly, rx, ry] of rectangle) {
    for (let x = lx * 2; x <= rx * 2; x++) {
      for (let y = ly * 2; y <= ry * 2; y++) {
        borders[x][y] = 1;
      }
    }
  }

  // 테두리 빼고 전부 지우기
  for (const [lx, ly, rx, ry] of rectangle) {
    for (let x = lx * 2 + 1; x < rx * 2; x++) {
      for (let y = ly * 2 + 1; y < ry * 2; y++) {
        borders[x][y] = 0;
      }
    }
  }
  return bfs(characterX * 2, characterY * 2);

  function bfs(startX, startY) {
    const queue = [[startX, startY, 0]];
    const visited = Array.from({ length: 102 }, () => Array(102).fill(0));
    let head = 0;

    visited[startX][startY] = 1;

    while (head < queue.length) {
      const [x, y, count] = queue[head++];
      if (x === itemX * 2 && y === itemY * 2) {
        return count / 2;
      }

      for (let i = 0; i < 4; i++) {
        const nx = x + dx[i];
        const ny = y + dy[i];

        if (nx < 0 || ny < 0 || nx >= 102 || ny >= 102) continue;
        if (visited[nx][ny]) continue;
        if (!borders[nx][ny]) continue;

        visited[nx][ny] = 1;
        queue.push([nx, ny, count + 1]);
      }
    }
  }
}
